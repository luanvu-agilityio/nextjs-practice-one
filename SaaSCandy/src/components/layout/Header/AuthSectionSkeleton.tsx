import { JSX } from 'react';

function AuthSectionSkeleton(): JSX.Element {
  return (
    <div className='hidden sm:flex items-center gap-4'>
      {/* Loading skeleton */}
      <div className='w-20 h-8 bg-gray-200 rounded animate-pulse'></div>
      <div className='w-20 h-8 bg-gray-200 rounded animate-pulse'></div>
    </div>
  );
}
export default AuthSectionSkeleton;
