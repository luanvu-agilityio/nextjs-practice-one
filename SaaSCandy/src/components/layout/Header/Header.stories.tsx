import type { Meta, StoryObj } from '@storybook/nextjs';
import { Header } from './index';

const meta: Meta<typeof Header> = {
  title: 'Layout/Header',
  component: Header,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;
type Story = StoryObj<typeof Header>;

export const Default: Story = {
  render: () => <Header />,
};

export const Mobile: Story = {
  parameters: {
    viewport: {
      defaultViewport: 'mobile1',
    },
  },
  render: () => <Header />,
};
