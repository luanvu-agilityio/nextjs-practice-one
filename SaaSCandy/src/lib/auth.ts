'use server';

import { signIn } from 'next-auth/react';

// API
import { authApi } from '@/api';

// Constants
import { AUTH_ROUTES } from '@/constants/auth-routes';

// Server Action for Sign In
export async function signInAction(formData: FormData) {
  const email = formData.get('email') as string;
  const password = formData.get('password') as string;

  try {
    const response = await authApi.login({ email, password });

    if (!response.user) {
      return {
        error: {
          status: 401,
          message: 'Invalid credentials',
        },
      };
    }

    return {
      success: true,
      user: response.user,
      credentials: { email, password },
    };
  } catch (error) {
    console.error('Sign in API error:', error);

    return { error };
  }
}

export async function signUpAction(formData: FormData) {
  const name = formData.get('name') as string;
  const email = formData.get('email') as string;
  const password = formData.get('password') as string;

  try {
    const { firstName, lastName } = extractNames(name);

    const response = await authApi.register({
      firstName,
      lastName,
      email,
      password,
    });

    return {
      success: true,
      user: response.user,
      credentials: { email, password },
    };
  } catch (error) {
    console.error('Sign up API error:', error);

    return { error };
  }
}

// Server Action for Social Sign In
export async function socialSignInAction(provider: string) {
  try {
    await signIn(provider, { callbackUrl: AUTH_ROUTES.HOME });
  } catch (error) {
    console.error(`Social sign in error (${provider}):`, error);
    return { error: `Failed to sign in with ${provider}` };
  }
}

// Helper function to extract first and last name
function extractNames(fullName: string) {
  const parts = fullName.trim().split(' ');
  const firstName = parts[0] || '';
  const lastName = parts.slice(1).join(' ') || '';
  return { firstName, lastName };
}
