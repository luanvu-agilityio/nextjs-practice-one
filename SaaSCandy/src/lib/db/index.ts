import { drizzle } from 'drizzle-orm/postgres-js';
import postgres from 'postgres';
import * as schema from './schema';

const connectionString = process.env.POSTGRES_URL!;
export const db = drizzle(postgres(connectionString), { schema });
export * from './schema';
