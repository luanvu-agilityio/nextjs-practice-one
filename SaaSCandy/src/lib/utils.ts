/**
 * Utility function for merging Tailwind CSS class names conditionally.
 *
 * - Uses `clsx` for conditional class logic.
 * - Uses `tailwind-merge` to resolve Tailwind class conflicts.
 *
 * @param inputs - List of class values (strings, arrays, objects).
 * @returns A single merged class name string.

 */
import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}
