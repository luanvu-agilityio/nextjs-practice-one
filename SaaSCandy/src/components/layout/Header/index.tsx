'use client';

import { JSX } from 'react';
import Link from 'next/link';
import { useSession } from 'next-auth/react';

// Icons
import LogoIcon from '@/components/icons/Logo';
import ThemeSwitcherIcon from '@/components/icons/ThemeSwitcherIcon';

// Components
import Navbar from '@/components/NavBar';
import { Button, Heading } from '@/components/common';
import UserMenu from '@/components/UserMenu';
import AuthSectionSkeleton from './AuthSectionSkeleton';

function Header(): JSX.Element {
  const { data: session, status } = useSession();

  const renderAuthSection = () => {
    if (status === 'loading') {
      return <AuthSectionSkeleton />;
    }

    if (session?.user) {
      return <UserMenu />;
    }

    return (
      <div className='hidden sm:flex items-center gap-4'>
        <Link href='/signin' className='no-underline'>
          <Button variant='secondary' size='small'>
            Sign In
          </Button>
        </Link>
        <Link href='/signup' className='no-underline'>
          <Button variant='primary' size='small'>
            Sign Up
          </Button>
        </Link>
      </div>
    );
  };

  return (
    <header className='bg-white w-full'>
      <div className='max-w-[1296px] mx-auto py-4 px-0 flex items-center gap-14'>
        <Link href='/' className='flex items-center gap-2 no-underline'>
          <LogoIcon className='w-8 h-8' />

          <div className='hidden sm:block'>
            <Heading
              as='h1'
              size='md'
              content={
                <span className='font-secondary text-primary text-3xl font-light'>
                  SaaS<span className='font-medium'>Candy</span>
                </span>
              }
            />
          </div>
        </Link>

        <Navbar />

        <div className='ml-auto flex items-center justify-between gap-4'>
          <ThemeSwitcherIcon className='w-10 h-10' />

          {/* Conditional rendering based on authentication status */}
          {renderAuthSection()}
        </div>
      </div>
    </header>
  );
}

export default Header;
