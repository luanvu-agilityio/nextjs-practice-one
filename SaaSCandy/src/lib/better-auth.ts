import { betterAuth } from 'better-auth';
import { drizzleAdapter } from 'better-auth/adapters/drizzle';
import { twoFactor } from 'better-auth/plugins';
import { Resend } from 'resend';
import { db } from './db';
import {
  ResetPasswordEmail,
  VerificationEmail,
} from '@/constants/email-template';

const resend = new Resend(process.env.RESEND_API_KEY);

export const auth = betterAuth({
  database: drizzleAdapter(db, {
    provider: 'pg',
  }),

  secret: process.env.BETTER_AUTH_SECRET!,
  baseURL: process.env.BETTER_AUTH_URL || 'http://localhost:3000',

  emailVerification: {
    enabled: true,
    sendOnSignUp: true,
    autoSignInAfterVerification: true,
    requireEmailVerification: true,
    sendVerificationEmail: async ({ user, token }) => {
      const verificationUrl = `${process.env.BETTER_AUTH_URL}/email-verification?token=${token}`;

      await resend.emails.send({
        from: process.env.RESEND_FROM_EMAIL || 'onboarding@resend.dev',
        to: user.email,
        subject: 'Verify your email - SaaSCandy',
        html: VerificationEmail({ verificationUrl }),
      });
    },
  },

  emailAndPassword: {
    enabled: true,
    requireEmailVerification: true,
    sendResetPassword: async ({ user, token }) => {
      const resetUrl = `${process.env.BETTER_AUTH_URL}/reset-password?token=${token}`;

      await resend.emails.send({
        from: process.env.RESEND_FROM_EMAIL || 'onboarding@resend.dev',
        to: user.email,
        subject: 'Reset your password - SaaSCandy',
        html: ResetPasswordEmail({ resetUrl }),
      });
    },
  },

  plugins: [
    twoFactor({
      issuer: 'SaaSCandy',
      otpOptions: {
        period: 30,
        digits: 6,
      },
    }),
  ],

  socialProviders: {
    google: {
      clientId: process.env.GOOGLE_CLIENT_ID || '',
      clientSecret: process.env.GOOGLE_CLIENT_SECRET || '',
      enabled: !!process.env.GOOGLE_CLIENT_ID,
    },
    github: {
      clientId: process.env.GITHUB_ID || '',
      clientSecret: process.env.GITHUB_SECRET || '',
      enabled: !!process.env.GITHUB_ID,
    },
  },

  logger: {
    level: 'debug',
  },
});

export type Session = typeof auth.$Infer.Session;
export type User = (typeof auth.$Infer.Session)['user'];
