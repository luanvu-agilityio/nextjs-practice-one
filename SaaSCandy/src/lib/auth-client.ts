/**
 * Client-side authentication utilities using Better Auth for React.
 *
 * - Provides sign-in, sign-up, sign-out, session hooks, and user update functions.
 * - Supports two-factor authentication (2FA) via the twoFactorClient plugin.
 * - Uses the app's base URL from environment variables.
 *
 * Usage:
 *   import { signIn, useSession } from '@/lib/auth-client';
 */
import { createAuthClient } from 'better-auth/react';
import { twoFactorClient } from 'better-auth/client/plugins';

export const authClient = createAuthClient({
  baseURL: process.env.NEXT_PUBLIC_APP_URL || 'http://localhost:3000',
  plugins: [twoFactorClient()],
});

export const {
  signIn,
  signUp,
  signOut,
  useSession,
  getSession,
  updateUser,
  twoFactor: { enable: enableTwoFactor, disable: disableTwoFactor, verifyOtp },
} = authClient;
