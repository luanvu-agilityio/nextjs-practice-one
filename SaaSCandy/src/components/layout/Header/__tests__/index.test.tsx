import { render, screen, fireEvent } from '@testing-library/react';
import Header from '@/components/layout/Header';
import { useSession } from '@/lib/auth-client';

jest.mock('@/lib/auth-client');
jest.mock('@/constants', () => ({
  ROUTES: {
    SIGN_IN: '/signin',
    SIGN_UP: '/signup',
  },
  NAV_LINKS: [
    { label: 'Home', href: '/' },
    { label: 'Services', href: '/services' },
    { label: 'About', href: '/about' },
    { label: 'Contact', href: '/contact' },
    { label: 'Blog', href: '/blog' },
  ],
}));

const mockUseSession = useSession as jest.MockedFunction<typeof useSession>;

describe('Header - Interactive Tests', () => {
  beforeEach(() => {
    return mockUseSession.mockReturnValue({
      data: null,
      isPending: false,
      error: null,
      refetch: function (): void {
        throw new Error('Function not implemented.');
      },
      isRefetching: false,
    });
  });

  it('should render logo and brand name', () => {
    render(<Header />);

    expect(screen.getByText(/SaaS/)).toBeInTheDocument();
    expect(screen.getByText(/Candy/)).toBeInTheDocument();
  });

  it('should render auth buttons when not authenticated', () => {
    render(<Header />);

    expect(screen.getByText('Sign In')).toBeInTheDocument();
    expect(screen.getByText('Sign Up')).toBeInTheDocument();
  });

  it('should show loading skeleton when session is pending', () => {
    mockUseSession.mockReturnValue({
      data: null,
      isPending: true,
      error: null,
      refetch: function (): void {
        throw new Error('Function not implemented.');
      },
      isRefetching: false,
    });

    const { container } = render(<Header />);

    const skeletons = container.querySelectorAll('.animate-pulse');
    expect(skeletons.length).toBeGreaterThan(0);
  });

  it('should toggle mobile auth menu', () => {
    render(<Header />);

    const toggleButton = screen.getByLabelText('Toggle auth menu');

    // Click to open
    fireEvent.click(toggleButton);

    // Mobile dropdown should appear (check for backdrop)
    const backdrop = document.querySelector('.fixed.inset-0');
    expect(backdrop).toBeInTheDocument();
  });

  it('should close mobile menu when backdrop clicked', () => {
    render(<Header />);

    const toggleButton = screen.getByLabelText('Toggle auth menu');
    fireEvent.click(toggleButton);

    const backdrop = document.querySelector('.fixed.inset-0') as HTMLElement;
    fireEvent.click(backdrop);

    // Menu should be closed
    expect(document.querySelector('.fixed.inset-0')).not.toBeInTheDocument();
  });
});
