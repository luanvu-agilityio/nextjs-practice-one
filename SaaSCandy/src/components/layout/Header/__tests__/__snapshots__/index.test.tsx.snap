// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Header - Snapshot matches snapshot 1`] = `
<div>
  <header
    class="bg-white w-full"
  >
    <div
      class="max-w-[1296px] mx-auto py-4 px-0 flex items-center gap-14"
    >
      <a
        href="/"
      >
        <svg
          class="w-8 h-8"
          data-testid="logo-icon"
        />
        <div
          class="hidden sm:block"
        >
          <h1>
            <span
              class="font-secondary text-primary text-3xl font-light"
            >
              SaaS
              <span
                class="font-medium"
              >
                Candy
              </span>
            </span>
          </h1>
        </div>
      </a>
      <nav
        data-testid="navbar"
      >
        Navbar
      </nav>
      <div
        class="ml-auto flex items-center justify-between gap-4"
      >
        <svg
          class="w-10 h-10"
          data-testid="theme-switcher-icon"
        />
        <div
          class="hidden sm:flex items-center gap-4"
        >
          <div
            class="w-20 h-8 bg-gray-200 rounded animate-pulse"
          />
          <div
            class="w-20 h-8 bg-gray-200 rounded animate-pulse"
          />
        </div>
      </div>
    </div>
  </header>
</div>
`;
